### PR TITLE
RPi: add clang & clang-check support

### PR DESCRIPTION
**Pull request by HidenoriKobayashi**
This adds support for clang & clang-check. clang emits more friendly warnings. clang-check does static checking of compiled codes. The checker requires cmake to generate compile command database (compile_commands.json). Some of the warnings generated are not harmful, but things like buffer overflow and memory leak might be worth a look. And it seems to me that some are actually bugs.

To use clang with this PR, you need to set a environment variable RPI_USE_CLANG=1 and make a symbolic link of clang and clang++ into where your cross gcc is. You also need a symbolic link to clang-check for the check.

### CODE REVIEW

File: Tools/clang-check.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (julianoes) [inline Tools/clang-check.sh]: Can you please add a comment here that this is specific for Raspberry Pi? Otherwise the file is a bit lost.

**Comment by julianoes** [comment]
@HidenoriKobayashi thanks for the PR, this is appreciated. I'll have a closer look soon.

File: Tools/clang-check.sh (hunk 1 above)
Comment (HidenoriKobayashi) [inline Tools/clang-check.sh]: Sure. Let me know if there is anything else that needs to be changed.

Comment (mcharleb) [inline Tools/clang-check.sh]: Perhaps OS/Platform specific tools should be in Firmware/Tools/<platform>/ to make it clear they are not intended for use on other platforms.

Comment (HidenoriKobayashi) [inline Tools/clang-check.sh]: @mcharleb That sounds good. So in this case, Firmware/Tools/rpi/clang-check.sh?
@julianoes Are you happy with that too?

Comment (julianoes) [inline Tools/clang-check.sh]: Sure, sounds good.

**Comment by HidenoriKobayashi** [comment]
Although there is no RPI specific clang, clang could not find correct assembler nor linker without this simlink. As I looked into the clang source code, it tried to find tools in two places: (1) where it (clang binary) is, and (2) where-it-finds-gcc-library-installation/../bin. The second one does not work because I'm abusing sysroot to point to where libc for RPI is, which is needed for the linker to find things like crt1.o but causes GCC installation detection to fail.

I guess changing the rpi_toolchains can fix the problem, but I thought this simple hack was enough for the purpose. If anyone can come up with better solution, I'm happy to throw away this one.

**Comment by LorenzMeier** [comment]
Can we generalise CLANG Check for all POSIX targets?

**Comment by dagar** [comment]
@HidenoriKobayashi could Tools/rpi/clang-check.sh move to a cmake add_custom_command? I think you have everything you need within cmake. Then we could even run clang-check in the builds.

**Comment by HidenoriKobayashi** [comment]
@LorenzMeier Theoretically, I believe it is even possible to perform the check for Nuttx targets. However, I've encountered many compile errors trying to do that. And for snapdragon, I need someone to help me as I don't have the hardware.

@dagar I will try to add a custom command that at least works for RPI. Allow me some more time please.

**Comment by HidenoriKobayashi** [comment]
@dagar  I take back what I've just said. clang-check takes too long for everyday build. And if we only occasionally run the check, I feel it's better to keep the check script outside cmake. If you still think it's better to have a cmake add_custom_command, could you please provide some suggestions how this should be done?

Besides, I guess I made a mistake about naming the file Toolchain-arm-linux-clang-gnueabihf-raspbian.cmake. Should this be renamed to something like Toolchain-arm-linux-gnueabihf-raspbian-clang.cmake?

**Comment by dagar** [comment]
I was hoping that moving it to cmake would be a relatively easy way to get it working for all targets.
Most of us use the posix_sitl_default build for local development and it can be built with clang. Would you be able to add clang-check for that target? I could set semaphore-ci to run it nightly.

**Comment by HidenoriKobayashi** [comment]
Then how about this one? I've extended the shell script to take arguments about the target and also to support clang-tidy. Now invoked with two arguments like this:
./Tools/clang-check.sh -b build_posix_sitl_default -t clang-tidy

**Comment by dagar** [comment]
That works for me. I'll give it a try tonight.

**Comment by dagar** [comment]
`./Tools/clang-tool.sh -b build_posix_sitl_default -t clang-tidy` worked for me. Once it's in master I'll take a closer look at the output. Most of it was unused compiler options and bad code in systemcmds/tests.

Does anyone object to the raspberry pi changes? @julianoes @mayanez

**Comment by HidenoriKobayashi** [comment]
Thanks for trying!

You may already know this, but just in case. clang-tidy checkers can be selectively disabled/enabled. You can also add a custom checks for PX4. Feel free to modify the shell script in any way you like.

**Comment by LorenzMeier** [comment]
Awesome! I've rebased and merged.

**Pull request closed without merging**